### PR TITLE
Add ability to analyse Regression Kink Analysis Designs

### LINKED ISSUE #223: Add ability to analyse Regression Kink Designs
* Card, D., Lee, D. S., Pei, Z., & Weber, A. (2017). Regression kink design: Theory and practice. In Regression discontinuity designs (Vol. 38, pp. 341-382). Emerald Publishing Limited.
* Card, D., Lee, D. S., Pei, Z., & Weber, A. (2015). Inference on causal effects in a generalized regression kink design. Econometrica, 83(6), 2453-2483.

### PR DESCRIPTION
**Pull request by drbenvincent**
This is my attempt to implement analysis methods for the Regression Kink Design. Happy to get any feedback on the work so far

- [X] First draft of notebook/docs
- [x] Implement `RegressionKink` class
  - [X] First attempt
  - [x] Add docstring for the `RegressionKink` class
  - [x] Add tests
  - [ ] ~~Refactor. At the moment this whole class is copied/pasted from the `RegressionDiscontinuity` class, so there is a lot of duplication. The main substantive difference is that we evaluate the function so as to evaluate the change in slopes either side of the kink.~~ Will create an issue about this because it gets into wider refactoring issues.
  - [x] Fix `summary` method
- [x] Add entry to Glossary + add reference back to glossary

Closes #223

**NOTE:** The example notebook does not try to act like an intro/explainer into regression kink designs. It assumes the reader has some background knowledge about what it is and when you'd use it. Happy to get input on whether it should have more introductory info, but there's a Discussion https://github.com/pymc-labs/CausalPy/discussions/262 on whether we should add a specific explainer section to the docs which would allow the notebooks to remain concise.

### CODE REVIEW

File: causalpy/pymc_experiments.py [hunk 1]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+            xi = np.linspace(
+                np.min(self.data[self.running_variable_name]),
+                np.max(self.data[self.running_variable_name]),
+                200,
+            )
+        self.x_pred = pd.DataFrame(
+            {self.running_variable_name: xi, "treated": self._is_treated(xi)}
+        )
+        # self.x_pred = pd.DataFrame({self.running_variable_name: xi})
+        (new_x,) = build_design_matrices([self._x_design_info], self.x_pred)
```
Comment (NathanielF) [inline causalpy/pymc_experiments.py:1042]: I'm sure there. is a reason but why is the output wrapped in braces?and then immediately into an array?

File: causalpy/tests/test_integration_pymc_examples.py [hunk 2]
```diff
@@ -217,6 +230,77 @@ def test_rd_drinking():
...
+    N = 50
+    kink = 0.5
+    beta = [0, -1, 0, 2, 0]
+    sigma = 0.05
+    # generate data
+    x = rng.uniform(-1, 1, N)
+    y = reg_kink_function(x, beta, kink) + rng.normal(0, sigma, N)
+    df = pd.DataFrame({"x": x, "y": y, "treated": x >= kink})
+    # run experiment
+    result = cp.pymc_experiments.RegressionKink(
```
Comment (NathanielF) [inline causalpy/tests/test_integration_pymc_examples.py:256]: Kind of neat that the kink model calls the linear model!

File: causalpy/tests/test_integration_pymc_examples.py [hunk 3]
```diff
@@ -217,6 +230,77 @@ def test_rd_drinking():
...
+        kink_point=kink,
+    )
+    assert isinstance(df, pd.DataFrame)
+    assert isinstance(result, cp.pymc_experiments.RegressionKink)
+    assert len(result.idata.posterior.coords["chain"]) == sample_kwargs["chains"]
+    assert len(result.idata.posterior.coords["draw"]) == sample_kwargs["draws"]
+
+
+@pytest.mark.integration
+def test_rkink_bandwidth():
```
Comment (NathanielF) [inline causalpy/tests/test_integration_pymc_examples.py:269]: I don't think you have an example in the notebook with the bandwidth parameter? Maybe worth adding/explaining

File: causalpy/pymc_experiments.py [hunk 4]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+    :param formula:
+        A statistical model formula
+    :param kink_point:
+        A scalar threshold value at which there is a change in the first derivative of
+        the assignment function
+    :param model:
+        A PyMC model
+    :param running_variable_name:
+        The name of the predictor variable that the kink_point is based upon
+    :param epsilon:
```
Comment (NathanielF) [inline causalpy/pymc_experiments.py:975]: I know you suggested not having the notebook as a place for explaining the theory of kink designs, but i feel like the differences between tweaking at least one of epsilon/bandwidth parameters could be mentioned or shown. 

It's fine if tweaking them isn't needed for your example, but it'd be good to hint at why they are there.

**Comment by NathanielF** [comment]
Notebook and experiment design look great. Added a few comments. Mostly nits.

**Review by NathanielF — APPROVED** [review]
Approved. Feel free to change/address the comments above as you like. Mostly they were small points of clarification. Major work looks great. Cool addition.

**Comment by drbenvincent** [comment]
<p>Done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/264/discussion/#comment-1793697850'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drbenvincent** [comment]
<p>As far as I can tell, when we use a quadratic, none of the individual parameters corresponds to the change in gradient at the kink point, so I've relied on calculating that numerically. But I have added a comment in about how the betas translate into different equations on the left and right of the kink point.</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/264/discussion/#comment-1793697863'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drbenvincent** [comment]
<p>Yep, I've removed the forest plot</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/264/discussion/#comment-1793697867'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drbenvincent** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/264/discussion/#comment-1793697869'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by drbenvincent** [comment]
<p>Sure. What I've actually done is to show 2 decimal places in the Arviz plot. At the moment, rounding to 1 decimal place in the plot method would require doing that for everything. So I'm just about to add an issue to add a round_to kwarg to the plot methods in causalpy.</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/264/discussion/#comment-1793697870'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: causalpy/pymc_experiments.py (hunk 1 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py:1042]: `build_design_matrices` returns both the x and y design matrices as a tuple, so we are just grabbing the x matrix here. And by default these are as dataframes. So you can either provide the `return_type='matrix'` argument to `build_design_matrices`, or manually do it yourself with `np.asarray` as I did here.

File: causalpy/tests/test_integration_pymc_examples.py (hunk 3 above)
Comment (drbenvincent) [reply causalpy/tests/test_integration_pymc_examples.py:269]: Good idea

File: causalpy/pymc_experiments.py (hunk 4 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py:975]: I've added an example to demonstrate use of the `bandwidth` parameter. And I've added an admonition box to explain what epsilon does.

File: causalpy/pymc_experiments.py [hunk 5]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+
+        self.gradient_left = (
+            self.pred_discon["posterior_predictive"].sel(obs_ind=1)["mu"]
+            - self.pred_discon["posterior_predictive"].sel(obs_ind=0)["mu"]
+        ) / self.epsilon
+        self.gradient_right = (
+            self.pred_discon["posterior_predictive"].sel(obs_ind=2)["mu"]
+            - self.pred_discon["posterior_predictive"].sel(obs_ind=1)["mu"]
+        ) / self.epsilon
+        self.gradient_change = self.gradient_right - self.gradient_left
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:1072]: I suggest modularizing this part in order to test individually each step. This to ensure (1) dode quality and (2) make sure we do not have edge cases

File: causalpy/pymc_experiments.py [hunk 6]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+        """
+        Plot the results
+        """
+        fig, ax = plt.subplots()
+        # Plot raw data
+        sns.scatterplot(
+            self.data,
+            x=self.running_variable_name,
+            y=self.outcome_variable_name,
+            c="k",  # hue="treated",
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:1100]: remove commented line

File: causalpy/pymc_experiments.py [hunk 7]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+            lw=3,
+            color="r",
+            label="treatment threshold",
+        )
+        ax.legend(
+            handles=(h_tuple for h_tuple in handles),
+            labels=labels,
+            fontsize=LEGEND_FONT_SIZE,
+        )
+        return (fig, ax)
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:1136]: you can remove the parentheses

File: causalpy/pymc_experiments.py [hunk 8]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+        else:
+            xi = np.linspace(
+                np.min(self.data[self.running_variable_name]),
+                np.max(self.data[self.running_variable_name]),
+                200,
+            )
+        self.x_pred = pd.DataFrame(
+            {self.running_variable_name: xi, "treated": self._is_treated(xi)}
+        )
+        # self.x_pred = pd.DataFrame({self.running_variable_name: xi})
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:1041]: remove un used line

File: causalpy/pymc_experiments.py [hunk 9]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+
+    def __init__(
+        self,
+        data: pd.DataFrame,
+        formula: str,
+        kink_point: float,
+        model=None,
+        running_variable_name: str = "x",
+        epsilon: float = 0.001,
+        bandwidth: Optional[float] = None,
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:991]: do you want to validate that `bandwidth> 0`?

File: causalpy/pymc_experiments.py [hunk 10]
```diff
@@ -957,6 +957,198 @@ def summary(self) -> None:
...
+        """
+        Print text output summarising the results
+        """
+
+        print(f"{self.expt_type:=^80}")
+        print(f"Formula: {self.formula}")
+        print(f"Running variable: {self.running_variable_name}")
+        print(f"Kink point on running variable: {self.kink_point}")
+        print("\nResults:")
+        print(f"Change in slope at kink point = {self.gradient_change.mean():.2f}")
```
Comment (juanitorduz) [inline causalpy/pymc_experiments.py:1148]: I think you can write this as a single print statement as 

```python
print(f"""
    ....
""")

**Review by juanitorduz — CHANGES_REQUESTED** [review]
some minor comments on the core component code style to make the computations more readable (and testable)

File: causalpy/pymc_experiments.py (hunk 7 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py]: done, 4 instances

File: causalpy/pymc_experiments.py (hunk 6 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py:1099]: done

File: causalpy/pymc_experiments.py (hunk 8 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py]: done

File: causalpy/pymc_experiments.py (hunk 10 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py]: Good idea

File: causalpy/pymc_experiments.py (hunk 9 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py]: We now have input validation that `bandwidth` and `epsilon` are >0 _and_ this is covered by tests that a `ValueError` exception is raised.

File: causalpy/pymc_experiments.py (hunk 5 above)
Comment (drbenvincent) [reply causalpy/pymc_experiments.py]: I've made a stab at this

**Comment by drbenvincent** [comment]
Thanks for the review @juanitorduz. I've made a stab at modularising the change in gradient calculation code and added some tests. Though I'm not sure if you it hits what you were thinking of.

File: causalpy/tests/test_input_validation.py [hunk 11]
```diff
@@ -223,3 +224,83 @@ def test_iv_treatment_var_is_present():
...
+# Regression kink design
+
+
+def setup_regression_kink_data(kink):
+    """Set up data for regression kink design tests"""
+    # define parameters for data generation
+    seed = 42
+    rng = np.random.default_rng(seed)
+    N = 50
+    kink = 0.5
```
Comment (juanitorduz) [inline causalpy/tests/test_input_validation.py:238]: Is this over-riding the king value passed to the function?

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Looks much nicer! thanks for the refactor. Letf a single comment :)

File: causalpy/tests/test_input_validation.py (hunk 11 above)
Comment (drbenvincent) [reply causalpy/tests/test_input_validation.py]: well spotted. Fixed

**Review by juanitorduz — APPROVED** [review]
Nice addition! Thank you @drbenvincent 🙌
